operators/src/pro/cache/shared_cache.rs

```diff
@@ -0,0 +1,1190 @@
+        // To avoid that, we just evict after we moved the entry from the landing zone to the cache.
+        // This is also not a problem since the total cache size is the cache size + the landing zone size.
+        self.evict_entries_until_can_fit_bytes(0);
+
+        Ok(cache_entry_id)
+    }
+}
+
+impl<T> Cache<RasterTile2D<T>> for CacheBackend
+where
+    T: Pixel + CacheElementSubType<CacheElementType = RasterTile2D<T>>,
+{
+    fn operator_cache_view_mut(
+        &mut self,
+        key: &CanonicOperatorName,
+    ) -> Option<OperatorCacheEntryView<RasterTile2D<T>>> {
+        self.raster_caches
+            .get_mut(key)
+            .map(|op| OperatorCacheEntryView {
+                operator_cache: op,
+                cache_size: &mut self.cache_size,
+                landing_zone_size: &mut self.landing_zone_size,
+                lru: &mut self.lru,
+            })
+    }
+}
+
+impl<T> Cache<FeatureCollection<T>> for CacheBackend
+where
+    T: Geometry + CacheElementSubType<CacheElementType = FeatureCollection<T>> + ArrowTyped,
+    FeatureCollection<T>: CacheElementHitCheck,
+{
+    fn operator_cache_view_mut(
+        &mut self,
+        key: &CanonicOperatorName,
+    ) -> Option<OperatorCacheEntryView<FeatureCollection<T>>> {
+        self.vector_caches
+            .get_mut(key)
+            .map(|op| OperatorCacheEntryView {
+                operator_cache: op,
+                cache_size: &mut self.cache_size,
+                landing_zone_size: &mut self.landing_zone_size,
+                lru: &mut self.lru,
+            })
+    }
+}
+
+impl CacheView<RasterCacheQueryEntry, RasterLandingQueryEntry> for CacheBackend {
+    fn operator_caches_mut(
+        &mut self,
+    ) -> &mut HashMap<CanonicOperatorName, RasterOperatorCacheEntry> {
+        &mut self.raster_caches
+    }
+}
+
+impl CacheView<VectorCacheQueryEntry, VectorLandingQueryEntry> for CacheBackend {
+    fn operator_caches_mut(
+        &mut self,
+    ) -> &mut HashMap<CanonicOperatorName, VectorOperatorCacheEntry> {
+        &mut self.vector_caches
+    }
+}
+
+pub trait CacheElement: ByteSize + Send + ByteSize
+where
+    Self: Sized,
+{
+    type Query: CacheQueryMatch + Clone + Send + Sync;
+    type LandingZoneContainer: LandingZoneElementsContainer<Self>;
+    type CacheContainer: CacheElementsContainer<Self::Query, Self, ResultStream = Self::ResultStream>
+        + From<Self::LandingZoneContainer>;
+    type ResultStream;
+    type CacheElementSubType: CacheElementSubType<CacheElementType = Self>;
+
+    fn move_into_landing_zone(
+        self,
+        landing_zone: &mut Self::LandingZoneContainer,
+    ) -> Result<(), CacheError> {
+        landing_zone.insert_element(self)
+    }
+
+    fn update_stored_query(&self, query: &mut Self::Query) -> Result<(), CacheError>;
+
+    fn cache_hint(&self) -> CacheHint;
+
+    fn typed_canonical_operator_name(key: CanonicOperatorName) -> TypedCanonicOperatorName;
+}
+
+pub trait CacheElementSubType {
```

Comment:
why can't this be implemented on the CacheElement?

Reply:
well thats a good question. I noticed that we need to specify the "SubType" u8...f64 for raster and point ... polygon for vector data when we use the cache.

with out that type as a trick we need to soemthing like  `let x = <cache as SharedCache<RasterTile>>:::remove(&mut cache, Id);` when using the "real" cache type "SharedCache". This is because when adding queries to the cache the cache does not know the type. 

When we implement it all on the CacheElement than we need to implement CacheElement for each of this subtypes. And then we also need to specify that type for all calls.